x/orderbook/msg.go

```diff
@@ -0,0 +1,104 @@
+package orderbook
+
+import (
+	"github.com/iov-one/weave"
+	coin "github.com/iov-one/weave/coin"
+	"github.com/iov-one/weave/errors"
+	"github.com/iov-one/weave/migration"
+)
+
+func init() {
+	// Migration needs to be registered for every message introduced in the codec.
+	// This is the convention to message versioning.
+	migration.MustRegister(1, &CreateOrderBookMsg{}, migration.NoModification)
+	migration.MustRegister(1, &CreateOrderMsg{}, migration.NoModification)
+	migration.MustRegister(1, &CancelOrderMsg{}, migration.NoModification)
+}
+
+const (
+	pathCreateOrderBook = "order/create_book"
+	pathCreateOrder     = "order/create"
+	pathCancelOrder     = "order/cancel"
+)
+
+var _ weave.Msg = (*CreateOrderBookMsg)(nil)
+var _ weave.Msg = (*CreateOrderMsg)(nil)
+var _ weave.Msg = (*CancelOrderMsg)(nil)
+
+// ROUTING, Path method fulfills weave.Msg interface to allow routing
+
+func (CreateOrderBookMsg) Path() string {
+	return pathCreateOrderBook
+}
+
+func (CreateOrderMsg) Path() string {
+	return pathCreateOrder
+}
+
+func (CancelOrderMsg) Path() string {
+	return pathCancelOrder
+}
+
+func (m CreateOrderBookMsg) Validate() error {
+	if err := m.Metadata.Validate(); err != nil {
+		return errors.Wrap(err, "metadata")
+	}
+	if err := validID(m.MarketID); err != nil {
+		return errors.Wrap(err, "market id")
+	}
+	if !coin.IsCC(m.AskTicker) {
+		return errors.Wrapf(errors.ErrCurrency, "Invalid Ask Ticker: %s", m.AskTicker)
+	}
+	if !coin.IsCC(m.BidTicker) {
+		return errors.Wrapf(errors.ErrCurrency, "Invalid Bid Ticker: %s", m.BidTicker)
+	}
+	if m.BidTicker <= m.AskTicker {
+		return errors.Wrapf(errors.ErrCurrency, "ask (%s) must be before bid (%s)", m.AskTicker, m.BidTicker)
+	}
+	return nil
+}
+
+func (m CreateOrderMsg) Validate() error {
+	if err := m.Metadata.Validate(); err != nil {
```

Comment:
Can you explain what ```m.Metadata.Validate()``` does in detail?

Reply:
Currently, it just Validates that Metadata is not nil and there is a positive integer schema set.

But in general, all structs used in public APIs have a `Validate() error` method. And when we embed one in another object, we can just Validate all the sub-structures and return the wrapped error, just giving the context why we were calling it (this makes more sense for a Coin which can be used in multiple fields, than a Metadata)